core/query/query.ml

```diff
@@ -713,179 +810,112 @@ struct
-            let x = Var.fresh_raw_var () in
-              (* Debug.print ("fresh variable: " ^ string_of_int x); *)
-              reduce_for_body ([(x, source)], [], body (Var (x, field_types)))
-          | v -> eval_error "Bad source in for comprehension: %s" (string_of_t v)
-  and reduce_for_body (gs, os, body) =
-    match body with
-      | Q.For (_, gs', os', body') -> Q.For (None, gs @ gs', os @ os', body')
-      | _                         -> Q.For (None, gs, os, body)
-  and reduce_if_condition (c, t, e) =
-    let open Q in
-    match c with
-      | Constant (Constant.Bool true) -> t
-      | Constant (Constant.Bool false) -> e
-      | If (c', t', _) ->
-        reduce_if_body
-          (reduce_or (reduce_and (c', t'),
-                      reduce_and (reduce_not c', t')),
-           t,
-           e)
-      | _ ->
-        if is_list t then
-          if e = nil then
-            reduce_where_then (c, t)
-          else
-            reduce_concat [reduce_where_then (c, t);
-                           reduce_where_then (reduce_not c, e)]
-        else
-          reduce_if_body (c, t, e)
-  and reduce_where_then (c, t) =
-    let open Q in
-    match t with
-      (* optimisation *)
-      | Constant (Constant.Bool true) -> t
-      | Constant (Constant.Bool false) -> Concat []
-
-      | Concat vs ->
-        reduce_concat (List.map (fun v -> reduce_where_then (c, v)) vs)
-      | For (_, gs, os, body) ->
-        For (None, gs, os, reduce_where_then (c, body))
-      | If (c', t', Concat []) ->
-        reduce_where_then (reduce_and (c, c'), t')
-      | _ ->
-        If (c, t, Concat [])
-  and reduce_if_body (c, t, e) =
-    let open Q in
-    match t with
-      | Record then_fields ->
-        begin match e with
-          | Record else_fields ->
-            assert (StringMap.equal (fun _ _ -> true) then_fields else_fields);
-            Record
-              (StringMap.fold
-                 (fun name t fields ->
-                   let e = StringMap.find name else_fields in
-                     StringMap.add name (reduce_if_body (c, t, e)) fields)
-                 then_fields
-                 StringMap.empty)
-          (* NOTE: this relies on any record variables having
-             been eta-expanded by this point *)
-          | _ -> eval_error "Mismatched fields"
+  and apply env : Q.t * Q.t list -> Q.t = function
+    | Q.Closure ((xs, body), closure_env), args ->
+      (* Debug.print ("Applying closure"); *)
+      (* Debug.print ("body: " ^ Ir.show_computation body); *)
+      (* Debug.print("Applying query closure: " ^ show_t (`Closure ((xs, body), closure_env))); *)
+      (* Debug.print("args: " ^ mapstrcat ", " show_t args); *)
+        let env = env ++ closure_env in
+        let env = List.fold_right2 (fun x arg env ->
+            bind env (x, arg)) xs args env in
+        (* Debug.print("Applied"); *)
+          norm_comp env body
+    | Q.Primitive "AsList", [xs] ->
```

Comment:
Perhaps this could be removed by `xlate` also

Reply:
Sounds ok.